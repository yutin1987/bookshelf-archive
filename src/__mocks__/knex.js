import knex from 'knex';
import { client } from 'jest-mock-knex';

export { client } from 'jest-mock-knex';

export default () => knex({ client });
