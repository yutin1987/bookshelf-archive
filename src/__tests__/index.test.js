import knex, { client } from 'jest-mock-knex';
import _ from 'lodash';
// import knex from 'knex';
import bookshelf from 'bookshelf';
import bookshelfArchive from '../';
import config from '../../knexfile';

process.setMaxListeners(0);

const sql = knex(config);
const db = bookshelf(sql);
db.plugin(bookshelfArchive);

const archive = ['more'];

class ClassModel extends db.Model {
  static tableName = 'class_table';
  static softDelete = false;
}

class StaticModel extends db.Model {
  static tableName = 'static_table';
  static hasTimestamps = true;
  static softDelete = true;
  static softField = 'disabled';
  static archive = archive;
  static archiveField = 'detail';
}

const ExtendModel = db.Model.extend({
  tableName: 'extend_table',
  softDelete: false,
});

const ProtoModel = db.Model.extend({
  tableName: 'proto_table',
  hasTimestamps: true,
  softDelete: true,
  softField: 'closed',
  archive,
});

const DefaultModel = db.Model.extend({
  tableName: 'default_table',
});

const data = {
  maxSize: 10,
  text: 'wow!! happy',
  more: { nickname: 'yutin', sex: true },
};

describe('bookshelf-archive', () => {
  beforeAll(async () => {
    await sql('class_table').truncate();
    await sql('static_table').truncate();
    await sql('extend_table').truncate();
    await sql('proto_table').truncate();
    await sql('default_table').truncate();
  });

  it('when insert', async () => {
    await _.reduce([{
      Model: ClassModel,
      query: {
        table: 'class_table',
        max_size: data.maxSize,
        text: data.text,
        more: JSON.stringify(data.more),
      },
    }, {
      Model: StaticModel,
      query: {
        table: 'static_table',
        max_size: data.maxSize,
        text: data.text,
        detail: JSON.stringify({ more: data.more }),
        created_at: expect.any(Date),
        updated_at: expect.any(Date),
      },
    }, {
      Model: ExtendModel,
      query: {
        table: 'extend_table',
        max_size: data.maxSize,
        text: data.text,
        more: JSON.stringify(data.more),
      },
    }, {
      Model: ProtoModel,
      query: {
        table: 'proto_table',
        max_size: data.maxSize,
        text: data.text,
        archive: JSON.stringify({ more: data.more }),
        created_at: expect.any(Date),
        updated_at: expect.any(Date),
      },
    }, {
      Model: DefaultModel,
      query: {
        table: 'default_table',
        max_size: data.maxSize,
        text: data.text,
        more: JSON.stringify(data.more),
      },
    }], (result, { Model, query }) => result.then(async () => {
      await result;
      client.mockClear();
      const model = await new Model({ ...data }).save();
      expect(client).toMatchSnapshot();
      expect(client).toHaveBeenCalledTimes(1);
      expect(client).toHaveBeenLastCalledWith(expect.objectContaining(query));
      expect(model.id).toBe('1');
    }), Promise.resolve());
  });

  it('when select', async () => {
    await _.reduce([{
      Model: ClassModel,
      reply: { ...data, id: '1', maxSize: `${data.maxSize}`, more: JSON.stringify(data.more) },
      query: { table: 'class_table' },
    }, {
      Model: StaticModel,
      reply: { ...data, id: '1', maxSize: `${data.maxSize}`, createdAt: expect.any(Date), updatedAt: expect.any(Date) },
      query: { table: 'static_table', disabled: 'NULL' },
    }, {
      Model: ExtendModel,
      reply: { ...data, id: '1', maxSize: `${data.maxSize}`, more: JSON.stringify(data.more) },
      query: { table: 'extend_table' },
    }, {
      Model: ProtoModel,
      reply: { ...data, id: '1', maxSize: `${data.maxSize}`, createdAt: expect.any(Date), updatedAt: expect.any(Date) },
      query: { table: 'proto_table', closed: 'NULL' },
    }, {
      Model: DefaultModel,
      reply: { ...data, id: '1', maxSize: `${data.maxSize}`, more: JSON.stringify(data.more) },
      query: { table: 'default_table', deleted_at: 'NULL' },
    }], (result, { Model, reply, query }) => result.then(async () => {
      await result;
      client.mockClear();
      const model = await (new Model()).fetch();
      expect(client).toMatchSnapshot();
      expect(model.toJSON()).toEqual(expect.objectContaining(reply));
      expect(client).toHaveBeenCalledTimes(1);
      expect(client).toHaveBeenLastCalledWith(expect.objectContaining(query));
    }), Promise.resolve());
  });

  it('when select with where', async () => {
    await _.reduce([{
      Model: ClassModel,
      reply: { ...data, id: '1', maxSize: `${data.maxSize}`, more: JSON.stringify(data.more) },
      query: { table: 'class_table', id: [1, 2], max_size: data.maxSize },
    }, {
      Model: StaticModel,
      reply: { ...data, id: '1', maxSize: `${data.maxSize}`, createdAt: expect.any(Date), updatedAt: expect.any(Date) },
      query: { table: 'static_table', id: [1, 2], disabled: 'NULL', max_size: data.maxSize },
    }, {
      Model: ExtendModel,
      reply: { ...data, id: '1', maxSize: `${data.maxSize}`, more: JSON.stringify(data.more) },
      query: { table: 'extend_table', id: [1, 2], max_size: data.maxSize },
    }, {
      Model: ProtoModel,
      reply: { ...data, id: '1', maxSize: `${data.maxSize}`, createdAt: expect.any(Date), updatedAt: expect.any(Date) },
      query: { table: 'proto_table', id: [1, 2], closed: 'NULL', max_size: data.maxSize },
    }, {
      Model: DefaultModel,
      reply: { ...data, id: '1', maxSize: `${data.maxSize}`, more: JSON.stringify(data.more) },
      query: { table: 'default_table', id: [1, 2], deleted_at: 'NULL', max_size: data.maxSize },
    }], (result, { Model, reply, query }) => result.then(async () => {
      client.mockClear();
      const model = new Model({ maxSize: data.maxSize });
      model.query('whereIn', 'id', [1, 2]);
      expect((await model.fetch()).toJSON()).toEqual(expect.objectContaining(reply));
      expect(client).toMatchSnapshot();
      expect(client).toHaveBeenCalledTimes(1);
      expect(client).toHaveBeenLastCalledWith(expect.objectContaining(query));
    }), Promise.resolve());
  });

  it('when select all', async () => {
    await _.reduce([{
      Model: ClassModel,
      reply: { ...data, id: '1', maxSize: `${data.maxSize}`, more: JSON.stringify(data.more) },
      query: { table: 'class_table' },
    }, {
      Model: StaticModel,
      reply: { ...data, id: '1', maxSize: `${data.maxSize}`, createdAt: expect.any(Date), updatedAt: expect.any(Date) },
      query: { table: 'static_table', disabled: 'NULL' },
    }, {
      Model: ExtendModel,
      reply: { ...data, id: '1', maxSize: `${data.maxSize}`, more: JSON.stringify(data.more) },
      query: { table: 'extend_table' },
    }, {
      Model: ProtoModel,
      reply: { ...data, id: '1', maxSize: `${data.maxSize}`, createdAt: expect.any(Date), updatedAt: expect.any(Date) },
      query: { table: 'proto_table', closed: 'NULL' },
    }, {
      Model: DefaultModel,
      reply: { ...data, id: '1', maxSize: `${data.maxSize}`, more: JSON.stringify(data.more) },
      query: { table: 'default_table', deleted_at: 'NULL' },
    }], (result, { Model, reply, query }) => result.then(async () => {
      client.mockClear();
      expect((await Model.fetchAll()).toJSON()[0]).toEqual(expect.objectContaining(reply));
      expect(client).toMatchSnapshot();
      expect(client).toHaveBeenCalledTimes(1);
      expect(client).toHaveBeenLastCalledWith(expect.objectContaining(query));
    }), Promise.resolve());
  });

  it('when select use query', async () => {
    await _.reduce([{
      Model: ClassModel,
      reply: { ...data, id: '1', maxSize: `${data.maxSize}`, more: JSON.stringify(data.more) },
      query: { table: 'class_table', max_size: data.maxSize },
    }, {
      Model: StaticModel,
      reply: { ...data, id: '1', maxSize: `${data.maxSize}`, createdAt: expect.any(Date), updatedAt: expect.any(Date) },
      query: { table: 'static_table', max_size: data.maxSize },
    }, {
      Model: ExtendModel,
      reply: { ...data, id: '1', maxSize: `${data.maxSize}`, more: JSON.stringify(data.more) },
      query: { table: 'extend_table', max_size: data.maxSize },
    }, {
      Model: ProtoModel,
      reply: { ...data, id: '1', maxSize: `${data.maxSize}`, createdAt: expect.any(Date), updatedAt: expect.any(Date) },
      query: { table: 'proto_table', max_size: data.maxSize },
    }, {
      Model: DefaultModel,
      reply: { ...data, id: '1', maxSize: `${data.maxSize}`, more: JSON.stringify(data.more) },
      query: { table: 'default_table', max_size: data.maxSize },
    }], (result, { Model, reply, query }) => result.then(async () => {
      client.mockClear();
      expect(
        (await Model.query().where('maxSize', data.maxSize).select('*'))[0],
      ).toEqual(expect.objectContaining(reply));
      expect(client).toMatchSnapshot();
      expect(client).toHaveBeenCalledTimes(1);
      expect(client).toHaveBeenLastCalledWith(expect.objectContaining(query));
    }), Promise.resolve());
  });

  it('when update', async () => {
    await _.reduce([{
      Model: ClassModel,
      query: {
        table: 'class_table',
        more: JSON.stringify({ nickname: 'orz' }),
      },
    }, {
      Model: StaticModel,
      query: {
        table: 'static_table',
        detail: JSON.stringify({ more: { nickname: 'orz' } }),
      },
    }, {
      Model: ExtendModel,
      query: {
        table: 'extend_table',
        more: JSON.stringify({ nickname: 'orz' }),
      },
    }, {
      Model: ProtoModel,
      query: {
        table: 'proto_table',
        archive: JSON.stringify({ more: { nickname: 'orz' } }),
      },
    }, {
      Model: DefaultModel,
      query: {
        table: 'default_table',
        more: JSON.stringify({ nickname: 'orz' }),
      },
    }], (result, { Model, query }) => result.then(async () => {
      client.mockClear();
      await new Model({ id: 1 }).save({ maxSize: 99, more: { nickname: 'orz' } });
      expect(client).toMatchSnapshot();
      expect(client).toHaveBeenCalledTimes(1);
      expect(client).toHaveBeenLastCalledWith(expect.objectContaining({ ...query, max_size: 99 }));
    }), Promise.resolve());
  });

  it('when destroy', async () => {
    await _.reduce([{
      Model: ClassModel,
      query: { table: 'class_table', method: 'delete' },
    }, {
      Model: StaticModel,
      query: { table: 'static_table', method: 'update', disabled: expect.any(Date) },
    }, {
      Model: ExtendModel,
      query: { table: 'extend_table', method: 'delete' },
    }, {
      Model: ProtoModel,
      query: { table: 'proto_table', method: 'update', closed: expect.any(Date) },
    }, {
      Model: DefaultModel,
      query: { table: 'default_table', method: 'update', deleted_at: expect.any(Date) },
    }], (result, { Model, query }) => result.then(async () => {
      client.mockClear();
      await new Model({ id: 1 }).destroy();
      expect(client).toMatchSnapshot();
      expect(client).toHaveBeenCalledTimes(1);
      expect(client).toHaveBeenLastCalledWith(expect.objectContaining(query));
    }), Promise.resolve());

    client.mockClear();
    let error;
    try {
      await new DefaultModel({ id: 1 }).destroy({ require: true });
    } catch (e) { error = e; }
    expect(client).toMatchSnapshot();
    expect(client).toHaveBeenCalledTimes(1);
    expect(error.message).toEqual('No Rows Deleted');
  });
});
