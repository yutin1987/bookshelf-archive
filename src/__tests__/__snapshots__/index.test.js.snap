// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bookshelf-archive when destroy 1`] = `
Array [
  "delete from class_table where id = 1",
]
`;

exports[`bookshelf-archive when destroy 2`] = `
Array [
  "update static_table set disabled = 'DATE' where id = 1",
]
`;

exports[`bookshelf-archive when destroy 3`] = `
Array [
  "delete from extend_table where id = 1",
]
`;

exports[`bookshelf-archive when destroy 4`] = `
Array [
  "update proto_table set closed = 'DATE' where id = 1",
]
`;

exports[`bookshelf-archive when destroy 5`] = `
Array [
  "update default_table set deleted_at = 'DATE' where id = 1",
]
`;

exports[`bookshelf-archive when destroy 6`] = `
Array [
  "update default_table set deleted_at = 'DATE' where id = 1",
]
`;

exports[`bookshelf-archive when insert 1`] = `
Array [
  "insert into class_table (max_size, more, text) values (10, '{nickname:yutin,sex:true}', 'wow!! happy') returning id",
]
`;

exports[`bookshelf-archive when insert 2`] = `
Array [
  "insert into static_table (created_at, detail, max_size, text, updated_at) values ('DATE', '{more:{nickname:yutin,sex:true}}', 10, 'wow!! happy', 'DATE') returning id",
]
`;

exports[`bookshelf-archive when insert 3`] = `
Array [
  "insert into extend_table (max_size, more, text) values (10, '{nickname:yutin,sex:true}', 'wow!! happy') returning id",
]
`;

exports[`bookshelf-archive when insert 4`] = `
Array [
  "insert into proto_table (archive, created_at, max_size, text, updated_at) values ('{more:{nickname:yutin,sex:true}}', 'DATE', 10, 'wow!! happy', 'DATE') returning id",
]
`;

exports[`bookshelf-archive when insert 5`] = `
Array [
  "insert into default_table (max_size, more, text) values (10, '{nickname:yutin,sex:true}', 'wow!! happy') returning id",
]
`;

exports[`bookshelf-archive when select 1`] = `
Array [
  "select class_table.* from class_table limit 1",
]
`;

exports[`bookshelf-archive when select 2`] = `
Array [
  "select static_table.* from static_table where static_table.disabled is null limit 1",
]
`;

exports[`bookshelf-archive when select 3`] = `
Array [
  "select extend_table.* from extend_table limit 1",
]
`;

exports[`bookshelf-archive when select 4`] = `
Array [
  "select proto_table.* from proto_table where proto_table.closed is null limit 1",
]
`;

exports[`bookshelf-archive when select 5`] = `
Array [
  "select default_table.* from default_table where default_table.deleted_at is null limit 1",
]
`;

exports[`bookshelf-archive when select all 1`] = `
Array [
  "select class_table.* from class_table",
]
`;

exports[`bookshelf-archive when select all 2`] = `
Array [
  "select static_table.* from static_table where static_table.disabled is null",
]
`;

exports[`bookshelf-archive when select all 3`] = `
Array [
  "select extend_table.* from extend_table",
]
`;

exports[`bookshelf-archive when select all 4`] = `
Array [
  "select proto_table.* from proto_table where proto_table.closed is null",
]
`;

exports[`bookshelf-archive when select all 5`] = `
Array [
  "select default_table.* from default_table where default_table.deleted_at is null",
]
`;

exports[`bookshelf-archive when select use query 1`] = `
Array [
  "select * from class_table where max_size = 10",
]
`;

exports[`bookshelf-archive when select use query 2`] = `
Array [
  "select * from static_table where max_size = 10",
]
`;

exports[`bookshelf-archive when select use query 3`] = `
Array [
  "select * from extend_table where max_size = 10",
]
`;

exports[`bookshelf-archive when select use query 4`] = `
Array [
  "select * from proto_table where max_size = 10",
]
`;

exports[`bookshelf-archive when select use query 5`] = `
Array [
  "select * from default_table where max_size = 10",
]
`;

exports[`bookshelf-archive when select with where 1`] = `
Array [
  "select class_table.* from class_table where id in (1, 2) and class_table.max_size = 10 limit 1",
]
`;

exports[`bookshelf-archive when select with where 2`] = `
Array [
  "select static_table.* from static_table where id in (1, 2) and static_table.max_size = 10 and static_table.disabled is null limit 1",
]
`;

exports[`bookshelf-archive when select with where 3`] = `
Array [
  "select extend_table.* from extend_table where id in (1, 2) and extend_table.max_size = 10 limit 1",
]
`;

exports[`bookshelf-archive when select with where 4`] = `
Array [
  "select proto_table.* from proto_table where id in (1, 2) and proto_table.max_size = 10 and proto_table.closed is null limit 1",
]
`;

exports[`bookshelf-archive when select with where 5`] = `
Array [
  "select default_table.* from default_table where id in (1, 2) and default_table.max_size = 10 and default_table.deleted_at is null limit 1",
]
`;

exports[`bookshelf-archive when update 1`] = `
Array [
  "update class_table set id = 1, max_size = 99, more = '{nickname:orz}' where id = 1",
]
`;

exports[`bookshelf-archive when update 2`] = `
Array [
  "update static_table set detail = '{more:{nickname:orz}}', id = 1, max_size = 99, updated_at = 'DATE' where id = 1",
]
`;

exports[`bookshelf-archive when update 3`] = `
Array [
  "update extend_table set id = 1, max_size = 99, more = '{nickname:orz}' where id = 1",
]
`;

exports[`bookshelf-archive when update 4`] = `
Array [
  "update proto_table set archive = '{more:{nickname:orz}}', id = 1, max_size = 99, updated_at = 'DATE' where id = 1",
]
`;

exports[`bookshelf-archive when update 5`] = `
Array [
  "update default_table set id = 1, max_size = 99, more = '{nickname:orz}' where id = 1",
]
`;
